File: backend/src/utils/fileValidation.ts
/**
 * Shared file validation utilities
 * Centralizes file type validation logic to prevent duplication
 */

export const ALLOWED_MIME_TYPES = [
  'image/jpeg',
  'image/jpg',
  'image/png',
  'application/pdf',
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet', // .xlsx
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document', // .docx
  'application/vnd.google-apps.document', // Google Docs (exported as PDF)
  'application/vnd.google-apps.spreadsheet', // Google Sheets (exported as XLSX)
] as const;

export const ALLOWED_EXTENSIONS = ['jpg', 'jpeg', 'png', 'pdf', 'xlsx', 'docx'] as const;

/**
 * Validate if a file type is allowed based on mime type and/or file extension
 */
export function isValidFileType(mimeType: string, fileName: string): boolean {
  const fileExtension = fileName.split('.').pop()?.toLowerCase();
  return (
    ALLOWED_MIME_TYPES.includes(mimeType as any) ||
    (fileExtension !== undefined && ALLOWED_EXTENSIONS.includes(fileExtension as any))
  );
}

/**
 * Get file extension from filename
 */
export function getFileExtension(fileName: string): string | undefined {
  return fileName.split('.').pop()?.toLowerCase();
}

/**
 * Validate file type and return error message if invalid
 */
export function validateFileType(mimeType: string, fileName: string): { valid: boolean; error?: string } {
  if (!isValidFileType(mimeType, fileName)) {
    return {
      valid: false,
      error: 'Invalid file type. Only JPG, PNG, PDF, XLSX, and DOCX files are allowed.',
    };
  }
  return { valid: true };
}

File: backend/utils/file_validation.py
"""
File validation utilities
"""
ALLOWED_MIME_TYPES = [
    'image/jpeg',
    'image/jpg',
    'image/png',
    'application/pdf',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',  # .xlsx
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',  # .docx
    'application/vnd.google-apps.document',  # Google Docs (exported as PDF)
    'application/vnd.google-apps.spreadsheet',  # Google Sheets (exported as XLSX)
]

ALLOWED_EXTENSIONS = ['jpg', 'jpeg', 'png', 'pdf', 'xlsx', 'docx']

def is_valid_file_type(mime_type: str, file_name: str) -> bool:
    """Check if file type is allowed"""
    file_extension = file_name.split('.')[-1].lower() if '.' in file_name else None
    return (
        mime_type in ALLOWED_MIME_TYPES or
        (file_extension and file_extension in ALLOWED_EXTENSIONS)
    )

def get_file_extension(file_name: str) -> str:
    """Get file extension from filename"""
    return file_name.split('.')[-1].lower() if '.' in file_name else None

def validate_file_type(mime_type: str, file_name: str) -> dict:
    """Validate file type and return error message if invalid"""
    if not is_valid_file_type(mime_type, file_name):
        return {
            'valid': False,
            'error': 'Invalid file type. Only JPG, PNG, PDF, XLSX, and DOCX files are allowed.'
        }
    return {'valid': True}

